File: internal/hub/messages.go
package hub

// InboundMessage is the envelope for all client→server JSON messages.
type InboundMessage struct {
	Event   string      `json:"event"`
	Payload interface{} `json:"payload"`
}

// -- Play room: client→server messages ---------------------------------------

// TeamRegisterMsg is sent by a new team client to register with a name.
type TeamRegisterMsg struct {
	TeamName string `json:"team_name"`
}

// TeamRejoinMsg is sent by a returning client to rejoin with their existing identity.
type TeamRejoinMsg struct {
	TeamID      string `json:"team_id"`
	DeviceToken string `json:"device_token"`
}

// DraftAnswerMsg is sent when a player drafts (but does not submit) an answer.
type DraftAnswerMsg struct {
	TeamName      string `json:"team_name"`
	RoundIndex    int    `json:"round_index"`
	QuestionIndex int    `json:"question_index"`
	Answer        string `json:"answer"`
}

// SubmitAnswersMsg is sent when a player locks in all answers for a round.
type SubmitAnswersMsg struct {
	TeamName   string        `json:"team_name"`
	RoundIndex int           `json:"round_index"`
	Answers    []AnswerEntry `json:"answers"`
}

// AnswerEntry is one answer within a SubmitAnswersMsg.
type AnswerEntry struct {
	QuestionIndex int    `json:"question_index"`
	Answer        string `json:"answer"`
}

// -- Host room: client→server messages ---------------------------------------

// HostLoadQuizMsg instructs the server to load a quiz from a file path.
type HostLoadQuizMsg struct {
	FilePath string `json:"file_path"`
}

// HostStartRoundMsg instructs the server to start the given round.
type HostStartRoundMsg struct {
	RoundIndex int `json:"round_index"`
}

// HostRevealQuestionMsg instructs the server to reveal a question.
type HostRevealQuestionMsg struct {
	RoundIndex    int `json:"round_index"`
	QuestionIndex int `json:"question_index"`
}

// HostMarkAnswerMsg instructs the server to mark a team's answer with a verdict.
type HostMarkAnswerMsg struct {
	TeamID        string `json:"team_id"`
	RoundIndex    int    `json:"round_index"`
	QuestionIndex int    `json:"question_index"`
	Verdict       string `json:"verdict"`
}

// HostCeremonyShowQuestionMsg instructs the server to show a ceremony question.
type HostCeremonyShowQuestionMsg struct {
	QuestionIndex int `json:"question_index"`
}

// HostCeremonyRevealAnswerMsg instructs the server to reveal the answer for a ceremony question.
type HostCeremonyRevealAnswerMsg struct {
	QuestionIndex int `json:"question_index"`
}

// HostPublishScoresMsg instructs the server to publish round scores.
type HostPublishScoresMsg struct {
	RoundIndex int `json:"round_index"`
}

// HostEndGameMsg instructs the server to end the game.
type HostEndGameMsg struct{}

File: frontend/src/ws/messages.ts
// Incoming messages (client -> server)

export interface TeamRegisterMsg {
  event: "team_register";
  payload: { team_name: string };
}

export interface TeamRejoinMsg {
  event: "team_rejoin";
  payload: { team_id: string; device_token: string };
}

export interface DraftAnswerMsg {
  event: "draft_answer";
  payload: {
    team_name: string;
    round_index: number;
    question_index: number;
    answer: string;
  };
}

export interface SubmitAnswersMsg {
  event: "submit_answers";
  payload: {
    team_id: string;
    round_index: number;
    answers: Array<{ question_index: number; answer: string }>;
  };
}

export interface HostLoadQuizMsg {
  event: "host_load_quiz";
  payload: { file_path: string };
}

export interface HostStartRoundMsg {
  event: "host_start_round";
  payload: { round_index: number };
}

export interface HostRevealQuestionMsg {
  event: "host_reveal_question";
  payload: { round_index: number; question_index: number };
}

export interface HostEndRoundMsg {
  event: "host_end_round";
  payload: { round_index: number };
}

export interface HostBeginScoringMsg {
  event: "host_begin_scoring";
  payload: { round_index: number };
}

export interface HostMarkAnswerMsg {
  event: "host_mark_answer";
  payload: {
    team_id: string;
    round_index: number;
    question_index: number;
    verdict: "correct" | "incorrect";
  };
}

export interface HostCeremonyShowQuestionMsg {
  event: "host_ceremony_show_question";
  payload: { question_index: number };
}

export interface HostCeremonyRevealAnswerMsg {
  event: "host_ceremony_reveal_answer";
  payload: { question_index: number };
}

export interface HostPublishScoresMsg {
  event: "host_publish_scores";
  payload: { round_index: number };
}

export interface HostEndGameMsg {
  event: "host_end_game";
  payload: Record<string, never>;
}

export type IncomingMessage =
  | TeamRegisterMsg
  | TeamRejoinMsg
  | DraftAnswerMsg
  | SubmitAnswersMsg
  | HostLoadQuizMsg
  | HostStartRoundMsg
  | HostRevealQuestionMsg
  | HostEndRoundMsg
  | HostBeginScoringMsg
  | HostMarkAnswerMsg
  | HostCeremonyShowQuestionMsg
  | HostCeremonyRevealAnswerMsg
  | HostPublishScoresMsg
  | HostEndGameMsg;

// Outgoing messages (server -> client)

export interface SessionCreatedMsg {
  event: "session_created";
  payload: { session_id: string };
}

export interface TeamRegisteredMsg {
  event: "team_registered";
  payload: { team_id: string; device_token: string };
}

export interface TeamJoinedMsg {
  event: "team_joined";
  payload: { team_id: string; team_name: string };
}

export interface QuizLoadedMsg {
  event: "quiz_loaded";
  payload: {
    title: string;
    round_count: number;
    question_count: number;
    player_url: string;
    display_url: string;
    confirmation: string;
    session_id: string;
  };
}

export interface RoundStartedMsg {
  event: "round_started";
  payload: { round_index: number; question_count: number };
}

export interface QuestionRevealedMsg {
  event: "question_revealed";
  payload: {
    question: { text: string; index: number };
    revealed_count: number;
    total_questions: number;
  };
}

export interface SubmissionAckMsg {
  event: "submission_ack";
  payload: { team_id: string; round_index: number; locked: boolean };
}

export interface SubmissionReceivedMsg {
  event: "submission_received";
  payload: { team_id: string; team_name: string; round_index: number };
}

export interface ScoringOpenedMsg {
  event: "scoring_opened";
  payload: { round_index: number };
}

export interface TeamSubmission {
  team_id: string;
  team_name: string;
  answer: string;
}

export interface ScoringQuestion {
  question_index: number;
  text: string;
  correct_answer: string;
  submissions: TeamSubmission[];
}

export interface ScoringDataMsg {
  event: "scoring_data";
  payload: { round_index: number; questions: ScoringQuestion[] };
}

export interface ScoreUpdatedMsg {
  event: "score_updated";
  payload: { team_id: string; round_index: number; running_total: number };
}

export interface CeremonyQuestionShownMsg {
  event: "ceremony_question_shown";
  payload: { question_index: number; question: { text: string; index: number } };
}

export interface CeremonyAnswerRevealedMsg {
  event: "ceremony_answer_revealed";
  payload: { question_index: number; answer: string };
}

export interface RoundScoresPublishedMsg {
  event: "round_scores_published";
  payload: { round_index: number; scores: Record<string, number> };
}

export interface GameOverMsg {
  event: "game_over";
  payload: { final_scores: Record<string, number> };
}

export interface StateSnapshotMsg {
  event: "state_snapshot";
  payload: Record<string, unknown>;
}

export interface ErrorMsg {
  event: "error";
  payload: { code: string; message: string };
}

export type OutgoingMessage =
  | SessionCreatedMsg
  | TeamRegisteredMsg
  | TeamJoinedMsg
  | QuizLoadedMsg
  | RoundStartedMsg
  | QuestionRevealedMsg
  | SubmissionAckMsg
  | SubmissionReceivedMsg
  | ScoringOpenedMsg
  | ScoringDataMsg
  | ScoreUpdatedMsg
  | CeremonyQuestionShownMsg
  | CeremonyAnswerRevealedMsg
  | RoundScoresPublishedMsg
  | GameOverMsg
  | StateSnapshotMsg
  | ErrorMsg;
